### ISSUE TITLE
Misleading error on tuple struct field with docstring

### ISSUE THREAD

**Issue by tbillington** [rust-lang/rust#122463]
I tried this code: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=48a7a316c9d4825be960974fd87a8d8d

```rust
struct Foo {
    a: Vec<(
        /// Docstring
        f32,
        f32,
    )>,
}
```

I expected to see this happen: Add docstring to field in tuple struct

Instead, this happened: Misleading second error "constant provided when a type was expected".

```
   Compiling playground v0.0.1 (/playground)
error: expected type, found doc comment `/// Docstring`
 --> src/lib.rs:3:9
  |
3 |         /// Docstring
  |         ^^^^^^^^^^^^^ expected type
  |
help: expressions must be enclosed in braces to be used as const generic arguments
  |
2 ~     a: Vec<{ (
3 |         /// Docstring
4 |         f32,
5 |         f32,
6 ~     ) }>,
  |

error[E0747]: constant provided when a type was expected
 --> src/lib.rs:2:12
  |
2 |       a: Vec<(
  |  ____________^
3 | |         /// Docstring
4 | |         f32,
5 | |         f32,
6 | |     )>,
  | |_____^
```

### Meta

`rustc --version --verbose`:
```
rustc 1.76.0 (07dca489a 2024-02-04)
binary: rustc
commit-hash: 07dca489ac2d933c78d3c5158e3f43beefeb02ce
commit-date: 2024-02-04
host: aarch64-apple-darwin
release: 1.76.0
LLVM version: 17.0.6
```

**Comment by tbillington** [comment]
As an aside, would docstrings in "nested" tuples like this be a feature request? Docstrings are supported in tuple structs as of https://github.com/rust-lang/rust/pull/87451.

**Comment by workingjubilee** [comment]
@tbillington This isn't a docstring on a tuple, this is a docstring on a generic. There is no "actual" tuple type there because it's effectively an argument to the `Vec`. If you want to document this type, you should probably separately declare and document its fields.

**Comment by workingjubilee** [comment]
This code works:
```rust
struct Foo {
    a: Vec<(
        // Comment
        f32,
        f32,
    )>,
}
```
So I suspect the error has something to do with the fact that a type in this position won't successfully have an attribute attached to it, because `///` desugars into
```rust
#[doc = ""]
```
I believe the diagnostics should be updated to inform people of this.

**Comment by ManiSalahmand** [comment]
@rustbot claim

I’d like to work on this. I reproduced the misleading follow-up E0747 locally and have a small parser recovery change plus a compiler diagnostic regression test. The test keeps the useful “expected type, found doc comment” diagnostic without also treating the tuple type as an unbraced const generic argument.
